src/Shabalov/Task-3/Widget.cpp

```diff
@@ -0,0 +1,193 @@
+                        QVector3D(z*(-half_width +  x * part_width), -half_width + y * part_width, z * half_width),
+                        QVector2D(0, 0), QVector3D(0.0, 0.0, z)));
+            }
+        }
+    }
+    for (auto x = -1; x <= 1; x += 2) {
+        for (auto z = 0; z < N; z++) {
+            for (auto y = 0; y < N; y++) {
+                vertexes.emplace_back(VertexData(
+                        QVector3D(x * half_width, -half_width + y * part_width, x*(-half_width + z * part_width)),
+                        QVector2D(0, 0), QVector3D(x, 0.0, 0.0)));
+            }
+        }
+    }
+    for (auto y = -1; y <= 1; y += 2) {
+        for (auto x = 0; x < N; x++) {
+            for (auto z = 0; z < N; z++) {
+                vertexes.emplace_back(VertexData(
+                        QVector3D(-half_width + x * part_width, y * half_width, y*(-half_width + z * part_width)),
+                        QVector2D(0, 0), QVector3D(0.0, y, 0.0)));
+            }
+        }
+    }
+
+    std::vector<GLuint> indexes;
+    indexes.reserve(36 * pow (N-1, 2));
+    for (int i = 0; i < 6 * N * N; i += N * N) {
+        for (int j = 0; j < (N-1) * (N-1); j += N) {
+            for (int k = 0; k < (N - 1); k++) {
+                indexes.emplace_back(i + j + k + N);
+                indexes.emplace_back(i + j + k + 0);
+                indexes.emplace_back(i + j + k + N + 1);
+                indexes.emplace_back(i + j + k + N + 1);
+                indexes.emplace_back(i + j + k + 0);
+                indexes.emplace_back(i + j + k + 1);
+            }
+        }
+    }
+
+    m_objects.append(std::make_shared<Object3D>(vertexes, indexes,
+                                                QImage(":/textures/cube.jpg")));
+
+}
+
+void Widget::mousePressEvent(QMouseEvent *event) {
+  if (event->buttons() == Qt::LeftButton) {
+    m_mousePosition = QVector2D(event->localPos());
+    event->accept();
+  }
+  QWidget::mousePressEvent(event);
+}
+
+void Widget::mouseMoveEvent(QMouseEvent *event) {
+  if (event->buttons() != Qt::LeftButton)
+    return;
+  QVector2D diff = QVector2D(event->localPos()) - m_mousePosition;
+  m_mousePosition = QVector2D(event->localPos());
+
+  auto angle = diff.length() / 2.0;
+
+  QVector3D axis = QVector3D(diff.y(), diff.x(), 0.0);
+
+  m_rotation = QQuaternion::fromAxisAndAngle(axis, angle) * m_rotation;
+  update();
+}
+
+void Widget::wheelEvent(QWheelEvent *event) {
+  if (event->angleDelta().x() > 0) {
+    m_z += 0.25f;
+  } else if (event->angleDelta().x() < 0) {
+    m_z -= 0.25f;
+  }
+  update();
+}
+
+void Widget::timerEvent(QTimerEvent *) {
+  for (int i = 0; i < m_objects.size(); i++) {
+    if (i % 2 == 0) {
+      m_objects[i]->rotate(
+          QQuaternion::fromAxisAndAngle(x_rot, y_rot, z_rot, angleObject));
+    } else {
+      m_objects[i]->rotate(
+          QQuaternion::fromAxisAndAngle(x_rot, y_rot, z_rot, -angleObject));
+    }
+  }
+  update();
+}
+
+void Widget::setXRotation(float angle) {
```

Comment:
```suggestion
void Widget::setXRotation(const float angle) {
```

и в остальных функциях тоже